src/uct/ib/ud/base/ud_ep.c

```diff
@@ -1006,9 +1107,16 @@ void uct_ud_ep_process_rx(uct_ud_iface_t *iface, uct_ud_neth_t *neth, unsigned b
         goto out;
     }
 
-    if (ucs_unlikely(!is_am && (neth->packet_type & UCT_UD_PACKET_FLAG_PUT))) {
-        /* TODO: remove once ucp implements put */
-        uct_ud_ep_rx_put(neth, byte_len);
+    if (ucs_unlikely(!is_am)) {
+        if (neth->packet_type & UCT_UD_PACKET_FLAG_PUT) {
+            /* TODO: remove once ucp implements put */
+            uct_ud_ep_rx_put(neth, byte_len);
+        } else if (neth->packet_type & UCT_UD_PACKET_FLAG_CTL) {
+            uct_ud_ep_rx_ctl(iface, ep, neth);
+        } else {
+            uct_ud_ep_process_unexp_packet(iface, ep, neth, byte_len);
+        }
```

Comment:
IMO better return control messages handling to line 992

Reply:
but we should add them to frag_list too